File: src/main/java/be/selske/aoc2023/Day07.java
package be.selske.aoc2023;

import be.selske.aoc2023.benchmark.Day;

import java.util.Arrays;
import java.util.Comparator;
import java.util.List;
import java.util.function.Function;

import static java.lang.Integer.parseInt;
import static java.util.Comparator.comparingInt;
import static java.util.stream.Collectors.groupingBy;

public class Day07 extends Day {

    public Day07() {
        super(7);
    }

    public static void main(String[] args) {
        new Day07()
                .example()
                .puzzle()
                .benchmark()
                .verifyPart1("251029473")
                .verifyPart2("251003917");
    }

    @Override
    protected void solve(ResultContainer results, String input, String parameter) {
        int part1 = solve(input, Hand::part1);
        int part2 = solve(input, Hand::part2);

        results.setPart1(part1);
        results.setPart2(part2);
    }

    private int solve(String input, Function<String, Hand> parser) {
        List<Hand> hands = input.lines()
                .map(parser)
                .sorted(Hand.COMPARATOR)
                .toList();

        int part1 = 0;
        for (int i = 0; i < hands.size(); i++) {
            Hand hand = hands.get(i);
            part1 += (i + 1) * hand.bid;
        }
        return part1;
    }

    private enum HandType {
        HIGH_CARD,
        ONE_PAIR,
        TWO_PAIR,
        THREE_OF_A_KIND,
        FULL_HOUSE,
        FOUR_OF_A_KIND,
        FIVE_OF_A_KIND,
    }

    private record Hand(int bid, HandType type, int[] cardValues) {

        public static final Comparator<Hand> COMPARATOR = comparingInt((Hand hand) -> hand.type.ordinal())
                .thenComparing((a,b) -> Arrays.compare(a.cardValues, b.cardValues));

        public static Hand part1(String input) {
            String[] split = input.split(" ");
            String value = split[0];
            int bid = parseInt(split[1]);
            HandType type = getHandType(value.chars().boxed().toList());
            var cardValues = parseValues(value, 11);

            return new Hand(bid, type, cardValues);
        }

        public static Hand part2(String input) {
            String[] split = input.split(" ");
            String value = split[0];
            int bid = parseInt(split[1]);
            HandType type = Hand.getHandType(value.chars().filter(c -> c != 'J').boxed().toList());
            var cardValues = parseValues(value, 0);

            return new Hand(bid, type, cardValues);
        }

        private static HandType getHandType(List<Integer> values) {
            var counts = values.stream().collect(groupingBy(Function.identity()));
            var maxCount = counts.values().stream().mapToInt(List::size).max().orElse(0);
            var jokers = 5 - values.size();

            HandType type;
            if (maxCount + jokers == 5) {
                type = HandType.FIVE_OF_A_KIND;
            } else if (maxCount + jokers == 4) {
                type = HandType.FOUR_OF_A_KIND;
            } else if (maxCount + jokers == 3) {
                var minCount = counts.values().stream().mapToInt(List::size).min().orElse(0);
                if (minCount == 2) {
                    type = HandType.FULL_HOUSE;
                } else {
                    type = HandType.THREE_OF_A_KIND;
                }
            } else if (maxCount + jokers == 2) {
                var numberOfPairs = counts.values().stream().filter(v -> v.size() == 2).count();
                if (numberOfPairs + jokers == 2) {
                    type = HandType.TWO_PAIR;
                } else {
                    type = HandType.ONE_PAIR;
                }
            } else {
                type = HandType.HIGH_CARD;
            }
            return type;
        }

        private static int[] parseValues(String input, int jValue) {
            return input.chars()
                    .map(i -> switch ((char) i) {
                        case 'A' -> 14;
                        case 'K' -> 13;
                        case 'Q' -> 12;
                        case 'J' -> jValue;
                        case 'T' -> 10;
                        default -> Character.digit(i, 10);
                    })
                    .toArray();
        }

    }

}

File: src/main/python/Day07.py
import timeit
from collections import Counter
from dataclasses import dataclass
from typing import Callable

FIVE_OF_A_KIND = 6
FOUR_OF_A_KIND = 5
FULL_HOUSE = 4
THREE_OF_A_KIND = 3
TWO_PAIRS = 2
ONE_PAIR = 1
HIGH_CARD = 0

# file = open("../resources/day7_example.txt")
file = open("../resources/day7.txt")

lines = file.readlines()

start = timeit.default_timer()


def parse_value(c, j_value: int) -> int:
    match c:
        case 'T':
            return 10
        case 'J':
            return j_value
        case 'Q':
            return 12
        case 'K':
            return 13
        case 'A':
            return 14
    return int(c)


def parse_values(name: str, j_value: int) -> list[int]:
    return [parse_value(c, j_value) for c in name]


def get_hand_type(values: list) -> int:
    counts = Counter(values)

    max_count = max(counts.values(), default=0)
    jokers = 5 - len(values)
    if max_count + jokers == 5:
        return FIVE_OF_A_KIND
    elif max_count + jokers == 4:
        return FOUR_OF_A_KIND
    elif max_count + jokers == 3:
        min_count = min(counts.values())
        if min_count == 2:
            return FULL_HOUSE
        else:
            return THREE_OF_A_KIND
    elif max_count + jokers == 2:
        number_of_pairs = len([count for count in counts.values() if count == 2])
        if number_of_pairs + jokers == 2:
            return TWO_PAIRS
        else:
            return ONE_PAIR
    else:
        return HIGH_CARD


@dataclass
class Hand:
    bid: int
    type: int
    values: list[int]


def solve(parser: Callable[[str], Hand]) -> int:
    hands = [parser(line) for line in lines]
    hands.sort(key=lambda h: (h.type, h.values), reverse=False)
    return sum([(i + 1) * hand.bid for i, hand in enumerate(hands)])


def parse_part1(line) -> Hand:
    split = line.split(' ')
    values = parse_values(split[0], 11)
    hand_type = get_hand_type([int(c) for c in values])
    return Hand(int(split[1]), hand_type, values)


def parse_part2(line) -> Hand:
    split = line.split(' ')
    values = parse_values(split[0], 0)
    hand_type = get_hand_type([int(c) for c in values if c != 0])
    return Hand(int(split[1]), hand_type, values)


part1 = solve(parse_part1)
part2 = solve(parse_part2)

print('Took ', (timeit.default_timer() - start) * 1000, 'ms')
print('Part 1: ', part1)
print('Part 2: ', part2)
